ONSdigital/eq-questionnaire-runner
# Grand Calculated Summary with repeating answers

### What is the context of this PR?
This PR extends one of the existing grand calculated summary schemas to include a sum of calculated summaries with dynamic answers, and list repeating block answers. It adds tests for this schema and makes the required fixes for it to work fully.

The only functional change required was navigation to the previous block from a list repeating question, as the `return_to_block_id` and `return_to` parameters were lists, which `ListRepeatingQuestion` did not support. So this has reused a method from `router.py` to get the correct url.

Swapped a few public methods that were unused outside of `Router` to private, to satisfy pylint.

### How to review
Read through the card and verify that the required test cases are included in the new schema. Check that the test coverage includes the cases mentioned in the card.

### Note
If an answer is used in multiple grand calculated summaries in different sections, the return to journey will take you through the section you are in until it is complete, and then return to the grand calculated summary you came from, so other sections using the answer may then be incomplete on the Hub. This is expected, and to reproduce it 

1. Complete sections 1-6
2. Enter section 6 and change the url to point you at grand-calculated-summary-5
3. Use the edit link for the first calculated summary
4. Use the edit link for the first answer
5. change it and keep pressing continue to confirm each CS/GCS until you are back on the hub
6. Section-3 will now be in progress as it was independent of that journey but has a grand calculated summary using the answer you just changed

### How to test functionally
Use the `test_grand_calculated_summary_repeating_answers` schema to test and in particular the `Utility bills` and `Streaming services` sections which have dynamic answers and repeating answers respectively. Check that adding and removing items from the lists behaves as expected, as does changing answers. The grand calculated summary should only be enabled once the dependent calculated summary sections are complete.

### Checklist
